File: Kalorie/Kalorie/Features/Dashboard/DashboardViewModel.swift
//
//  DashboardViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 05.06.2024.
//

import Foundation
import MacroKit

struct DailyMacros {
    let calories: Int
    let protein: Double
    let carbs: Double
    let carbohydrateSugar: Double
    let fat: Double
    let fatUnsaturated: Double
    let fiber: Double
    let salt: Double

    init(
        calories: Int,
        protein: Double,
        carbs: Double,
        carbohydrateSugar: Double,
        fat: Double,
        fatUnsaturated: Double,
        fiber: Double,
        salt: Double
    ) {
        self.calories = calories
        self.protein = protein
        self.carbs = carbs
        self.carbohydrateSugar = carbohydrateSugar
        self.fat = fat
        self.fatUnsaturated = fatUnsaturated
        self.fiber = fiber
        self.salt = salt
    }

    init(foods: [FoodConsumedDomain]) {
        let macros = foods.map {
            Macros(
                calories: Int32($0.calories),
                protein: $0.protein,
                carbohydrate: $0.carbohydrate,
                carbohydrateSugar: $0.carbohydrateSugar,
                fat: $0.fat,
                fatUnsaturated: $0.fatUnsaturated,
                fiber: $0.fiber ?? 0,
                salt: $0.salt
            )
        }
        let total = MacrosKt.total(macros)
        calories = Int(total.calories)
        protein = total.protein
        carbs = total.carbohydrate
        carbohydrateSugar = total.carbohydrateSugar
        fat = total.fat
        fatUnsaturated = total.fatUnsaturated
        fiber = total.fiber
        salt = total.salt
    }
}

final class DashboardViewModel: ObservableObject {

    // MARK: - Properties

    @Published private(set) var state: LoadingState<Void> = .loading
    @Published var mealTypes: [MealTypeDomain] = []
    @Published var foodsConsumed: [FoodConsumedDomain] = []
    @Published var selectedDay = Date.now
    @Published var showMealTypeSheet = false
    @Published var showAddFoodSheet = false
    @Published var showCalendarSheet = false
    @Published var showAccountSheet = false
    @Published var showMyCreatedMealEditor = false
    @Published var alertItem: AlertItem?
    @Published var isDeleteConfirmationVisible = false
    @Published private(set) var activeDaysInMonth: Set<Int> = []

    private var isMyCreatedMealEditorPending = false
    private var isViewingToday = true
    private var hasCompletedInitialLoad = false
    private var monthCache: [String: [FoodConsumedDomain]] = [:]
    private var cachedMonthKeys: Set<String> = []
    private var foodPendingDeletion: FoodConsumedDomain?
    private var dayChangeObserver: NSObjectProtocol?

    private let fetchMealTypes: any FetchMealTypesUseCaseProtocol
    private let fetchFoodsConsumedForMonth: any FetchFoodsConsumedForMonthUseCaseProtocol
    private let setupDefaultMeals: any SetupDefaultMealsUseCaseProtocol
    private let confirmMealTypesEmpty: any ConfirmMealTypesEmptyUseCaseProtocol
    private let deleteFoodConsumed: any DeleteFoodConsumedUseCaseProtocol

    // MARK: - Init

    init(
        fetchMealTypes: any FetchMealTypesUseCaseProtocol,
        fetchFoodsConsumedForMonth: any FetchFoodsConsumedForMonthUseCaseProtocol,
        setupDefaultMeals: any SetupDefaultMealsUseCaseProtocol,
        confirmMealTypesEmpty: any ConfirmMealTypesEmptyUseCaseProtocol,
        deleteFoodConsumed: any DeleteFoodConsumedUseCaseProtocol
    ) {
        self.fetchMealTypes = fetchMealTypes
        self.fetchFoodsConsumedForMonth = fetchFoodsConsumedForMonth
        self.setupDefaultMeals = setupDefaultMeals
        self.confirmMealTypesEmpty = confirmMealTypesEmpty
        self.deleteFoodConsumed = deleteFoodConsumed
        observeDayChange()
    }

    deinit {
        if let dayChangeObserver {
            NotificationCenter.default.removeObserver(dayChangeObserver)
        }
    }

    // MARK: - Functions

    var dailyMacros: DailyMacros { DailyMacros(foods: foodsConsumed) }

    var groupedFoods: [(mealType: MealTypeDomain?, foods: [FoodConsumedDomain])] {
        var foodsByMealTypeId: [String: [FoodConsumedDomain]] = [:]
        var unassigned: [FoodConsumedDomain] = []

        for food in foodsConsumed {
            if let resolvedMealTypeId = mealTypes.resolvedMealTypeId(for: food) {
                foodsByMealTypeId[resolvedMealTypeId, default: []].append(food)
            } else {
                unassigned.append(food)
            }
        }

        var result: [(mealType: MealTypeDomain?, foods: [FoodConsumedDomain])] = mealTypes
            .sorted { $0.startTime < $1.startTime }
            .compactMap { mealType in
                foodsByMealTypeId[mealType.id].map { (mealType: mealType, foods: $0) }
            }
        if !unassigned.isEmpty {
            result.append((mealType: nil, foods: unassigned))
        }

        return result
    }

    @MainActor
    func onAppear() async {
        guard !hasCompletedInitialLoad else { return }
        selectedDay = Date.now
        isViewingToday = true
        state = .loading
        do {
            try await refreshMealTypes()
            try await loadMonth(for: selectedDay)
            foodsConsumed = foodsFromCache(for: selectedDay)
            state = .loaded
        } catch {
            Log.error(error, category: Constants.LogCategory.dashboard)
            alertItem = unknownErrorAlertItem(for: error)
            state = .loaded
        }
        hasCompletedInitialLoad = true
    }

    @MainActor
    func onRefresh() async {
        guard hasCompletedInitialLoad else { return }
        do {
            advanceSelectedDayIfNeeded()
            try await refreshMealTypes()
            invalidateCache(for: selectedDay)
            try await loadMonth(for: selectedDay)
            foodsConsumed = foodsFromCache(for: selectedDay)
        } catch {
            Log.error(error, category: Constants.LogCategory.dashboard)
            alertItem = unknownErrorAlertItem(for: error)
        }
    }

    @MainActor
    func onFoodConsumedUpdated() async {
        do {
            invalidateCache(for: selectedDay)
            try await loadMonth(for: selectedDay)
            foodsConsumed = foodsFromCache(for: selectedDay)
        } catch {
            Log.error(error, category: Constants.LogCategory.dashboard)
            alertItem = unknownErrorAlertItem(for: error)
        }
    }

    func onCreateMealRequested() {
        isMyCreatedMealEditorPending = true
    }

    func onDeleteRequested(_ food: FoodConsumedDomain) {
        foodPendingDeletion = food
        isDeleteConfirmationVisible = true
    }

    @MainActor
    func onDeleteConfirmed() async {
        guard let food = foodPendingDeletion else { return }
        foodPendingDeletion = nil
        do {
            try await deleteFoodConsumed(id: food.id)
            invalidateCache(for: selectedDay)
            try await loadMonth(for: selectedDay)
            foodsConsumed = foodsFromCache(for: selectedDay)
        } catch {
            Log.error(error, category: Constants.LogCategory.dashboard)
            alertItem = AlertItem(title: L10n.Dashboard.errorDeleteFailed)
        }
    }

    func onAddFoodSheetDismissed() {
        guard isMyCreatedMealEditorPending else { return }
        isMyCreatedMealEditorPending = false
        showMyCreatedMealEditor = true
    }

    @MainActor
    func onMealTypesChanged() async {
        do {
            try await refreshMealTypes()
        } catch {
            Log.error(error, category: Constants.LogCategory.dashboard)
            alertItem = unknownErrorAlertItem(for: error)
        }
    }

    @MainActor
    func onDayChanged(_ date: Date) async {
        isViewingToday = Calendar.current.isDate(date, inSameDayAs: Date.now)
        await loadFoods(for: date)
    }

    @MainActor
    func onDaySelected(_ date: Date) async {
        isViewingToday = Calendar.current.isDate(date, inSameDayAs: Date.now)
        selectedDay = date
        showCalendarSheet = false
        await loadFoods(for: date)
    }

    @MainActor
    func onCalendarMonthChanged(to month: Date) async {
        let key = monthCacheKey(for: month)
        if cachedMonthKeys.contains(key) {
            activeDaysInMonth = computeActiveDays(for: month)
        } else {
            do {
                try await loadMonth(for: month)
            } catch {
                Log.error(error, category: Constants.LogCategory.dashboard)
                alertItem = unknownErrorAlertItem(for: error)
            }
        }
    }

    // MARK: - Private

    private func isOffline(_ error: Error) -> Bool {
        (error as? FirestoreDataProviderError) == .unreachable
    }

    private func unknownErrorAlertItem(for error: Error) -> AlertItem {
        if isOffline(error) {
            AlertItem(title: L10n.Common.errorOffline, message: L10n.Common.errorOfflineMessage)
        } else {
            AlertItem(title: L10n.Common.errorUnknown, message: L10n.Common.errorUnknownMessage)
        }
    }

    @MainActor
    private func refreshMealTypes() async throws {
        var types = try await fetchMealTypes()
        if
            types.isEmpty,
            try await confirmMealTypesEmpty()
        {
            types = try await setupDefaultMeals()
        }
        mealTypes = types
    }

    private func advanceSelectedDayIfNeeded() {
        guard isViewingToday else { return }
        guard !Calendar.current.isDate(selectedDay, inSameDayAs: Date.now) else { return }
        selectedDay = Date.now
    }

    private func observeDayChange() {
        dayChangeObserver = NotificationCenter.default.addObserver(
            forName: .NSCalendarDayChanged,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            Task { @MainActor in await self?.onRefresh() }
        }
    }

    @MainActor
    private func loadFoods(for date: Date) async {
        let key = monthCacheKey(for: date)
        if cachedMonthKeys.contains(key) {
            foodsConsumed = foodsFromCache(for: date)
            activeDaysInMonth = computeActiveDays(for: date)
        } else {
            do {
                try await loadMonth(for: date)
                foodsConsumed = foodsFromCache(for: date)
            } catch {
                Log.error(error, category: Constants.LogCategory.dashboard)
                alertItem = unknownErrorAlertItem(for: error)
            }
        }
    }

    @MainActor
    private func loadMonth(for date: Date) async throws {
        let foods = try await fetchFoodsConsumedForMonth(for: date)
        populateCache(with: foods, for: date)
        activeDaysInMonth = computeActiveDays(for: date)
    }

    @MainActor
    private func monthCacheKey(for date: Date) -> String {
        date.formatCacheKey(with: "yyyy-MM")
    }

    @MainActor
    private func dayCacheKey(for date: Date) -> String {
        date.formatCacheKey(with: "yyyy-MM-dd")
    }

    @MainActor
    private func populateCache(with foods: [FoodConsumedDomain], for month: Date) {
        let key = monthCacheKey(for: month)
        monthCache = monthCache.filter { !$0.key.hasPrefix(key) }
        cachedMonthKeys.insert(key)
        for food in foods {
            let dayKey = dayCacheKey(for: food.date)
            monthCache[dayKey, default: []].append(food)
        }
    }

    @MainActor
    private func foodsFromCache(for date: Date) -> [FoodConsumedDomain] {
        monthCache[dayCacheKey(for: date)] ?? []
    }

    @MainActor
    private func computeActiveDays(for month: Date) -> Set<Int> {
        let key = monthCacheKey(for: month)
        var result = Set<Int>()
        for (dayKey, foods) in monthCache where dayKey.hasPrefix(key) && !foods.isEmpty {
            let parts = dayKey.split(separator: "-")
            if parts.count == 3, let day = Int(parts[2]) {
                result.insert(day)
            }
        }
        return result
    }

    @MainActor
    private func invalidateCache(for date: Date) {
        let key = monthCacheKey(for: date)
        monthCache = monthCache.filter { !$0.key.hasPrefix(key) }
        cachedMonthKeys.remove(key)
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/dashboard/DashboardViewModel.kt
package antoni.kalorie.features.dashboard

import androidx.compose.runtime.mutableStateListOf
import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.core.models.FoodConsumedDomain
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.models.resolvedMealTypeId
import antoni.kalorie.core.usecases.ConfirmMealTypesEmptyUseCaseProtocol
import antoni.kalorie.core.usecases.DeleteFoodConsumedUseCaseProtocol
import antoni.kalorie.core.usecases.FetchFoodsConsumedForMonthUseCaseProtocol
import antoni.kalorie.core.usecases.FetchMealTypesUseCaseProtocol
import antoni.kalorie.core.usecases.SetupDefaultMealsUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.LoadingState
import antoni.kalorie.core.utils.Log
import antoni.kalorie.core.utils.formatCacheKey
import antoni.kalorie.core.utils.isFirestoreUnreachable
import antoni.kalorie.core.utils.isSameDay
import antoni.kalorie.macrokit.Macros
import antoni.kalorie.macrokit.total
import java.time.Instant
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

data class DailyMacros(
    val calories: Int,
    val protein: Double,
    val carbs: Double,
    val carbohydrateSugar: Double,
    val fat: Double,
    val fatUnsaturated: Double,
    val fiber: Double,
    val salt: Double,
) {
    constructor(foods: List<FoodConsumedDomain>) : this(
        foods
            .map {
                Macros(
                    calories = it.calories,
                    protein = it.protein,
                    carbohydrate = it.carbohydrate,
                    carbohydrateSugar = it.carbohydrateSugar,
                    fat = it.fat,
                    fatUnsaturated = it.fatUnsaturated,
                    fiber = it.fiber ?: 0.0,
                    salt = it.salt,
                )
            }
            .total()
    )

    private constructor(total: Macros) : this(
        calories = total.calories,
        protein = total.protein,
        carbs = total.carbohydrate,
        carbohydrateSugar = total.carbohydrateSugar,
        fat = total.fat,
        fatUnsaturated = total.fatUnsaturated,
        fiber = total.fiber,
        salt = total.salt,
    )
}

data class FoodGroup(
    val mealType: MealTypeDomain?,
    val foods: List<FoodConsumedDomain>,
)

class DashboardViewModel(
    private val fetchMealTypes: FetchMealTypesUseCaseProtocol,
    private val fetchFoodsConsumedForMonth: FetchFoodsConsumedForMonthUseCaseProtocol,
    private val setupDefaultMeals: SetupDefaultMealsUseCaseProtocol,
    private val confirmMealTypesEmpty: ConfirmMealTypesEmptyUseCaseProtocol,
    private val deleteFoodConsumed: DeleteFoodConsumedUseCaseProtocol,
) : ViewModel() {

    // MARK: - Properties

    private val _state = MutableStateFlow<LoadingState<Unit>>(LoadingState.Loading)
    val state: StateFlow<LoadingState<Unit>> = _state
    val mealTypes = MutableStateFlow<List<MealTypeDomain>>(emptyList())
    val foodsConsumed = MutableStateFlow<List<FoodConsumedDomain>>(emptyList())
    val selectedDay = MutableStateFlow(Instant.now())
    val showMealTypeSheet = MutableStateFlow(false)
    val showAddFoodSheet = MutableStateFlow(false)
    val showCalendarSheet = MutableStateFlow(false)
    val showAccountSheet = MutableStateFlow(false)
    val alertItem = MutableStateFlow<AlertItem?>(null)
    val backStack = mutableStateListOf<DashboardDestination>(DashboardDestination.Dashboard)
    val isDeleteConfirmationVisible = MutableStateFlow(false)
    private val _activeDaysInMonth = MutableStateFlow<Set<Int>>(emptySet())
    val activeDaysInMonth: StateFlow<Set<Int>> = _activeDaysInMonth

    private var isViewingToday = true
    private var hasCompletedInitialLoad = false
    private var monthCache = mutableMapOf<String, List<FoodConsumedDomain>>()
    private val cachedMonthKeys = mutableSetOf<String>()
    private var foodPendingDeletion: FoodConsumedDomain? = null

    val dailyMacros: DailyMacros
        get() = DailyMacros(foodsConsumed.value)

    val groupedFoods: List<FoodGroup>
        get() {
            val foodsByMealTypeId = mutableMapOf<String, MutableList<FoodConsumedDomain>>()
            val unassigned = mutableListOf<FoodConsumedDomain>()

            for (food in foodsConsumed.value) {
                val resolvedMealTypeId = mealTypes.value.resolvedMealTypeId(food)
                if (resolvedMealTypeId != null) {
                    foodsByMealTypeId.getOrPut(resolvedMealTypeId) { mutableListOf() } += food
                } else {
                    unassigned += food
                }
            }

            val result = mealTypes.value
                .sortedBy { it.startMinutes }
                .mapNotNull { mealType -> foodsByMealTypeId[mealType.id]?.let { FoodGroup(mealType, it) } }
                .toMutableList()
            if (unassigned.isNotEmpty()) {
                result += FoodGroup(mealType = null, foods = unassigned)
            }
            return result
        }

    // MARK: - Functions

    suspend fun onAppear() {
        if (hasCompletedInitialLoad) return
        selectedDay.value = Instant.now()
        isViewingToday = true
        _state.value = LoadingState.Loading
        perform {
            refreshMealTypes()
            loadMonth(selectedDay.value)
            foodsConsumed.value = foodsFromCache(selectedDay.value)
        }
        _state.value = LoadingState.loaded
        hasCompletedInitialLoad = true
    }

    suspend fun onRefresh() {
        if (!hasCompletedInitialLoad) return
        perform {
            advanceSelectedDayIfNeeded()
            refreshMealTypes()
            invalidateCache(selectedDay.value)
            loadMonth(selectedDay.value)
            foodsConsumed.value = foodsFromCache(selectedDay.value)
        }
    }

    suspend fun onFoodConsumedUpdated() {
        perform {
            invalidateCache(selectedDay.value)
            loadMonth(selectedDay.value)
            foodsConsumed.value = foodsFromCache(selectedDay.value)
        }
    }

    fun onDeleteRequested(food: FoodConsumedDomain) {
        foodPendingDeletion = food
        isDeleteConfirmationVisible.value = true
    }

    suspend fun onDeleteConfirmed() {
        val food = foodPendingDeletion ?: return
        foodPendingDeletion = null
        perform(onFailure = { alertItem.value = AlertItem(titleRes = R.string.dashboard_error_deleteFailed) }) {
            deleteFoodConsumed(id = food.id)
            invalidateCache(selectedDay.value)
            loadMonth(selectedDay.value)
            foodsConsumed.value = foodsFromCache(selectedDay.value)
        }
    }

    suspend fun onMealTypesChanged() {
        perform { refreshMealTypes() }
    }

    suspend fun onDayChanged(date: Instant) {
        isViewingToday = date.isSameDay(Instant.now())
        loadFoods(date)
    }

    suspend fun onDaySelected(date: Instant) {
        isViewingToday = date.isSameDay(Instant.now())
        selectedDay.value = date
        showCalendarSheet.value = false
        loadFoods(date)
    }

    suspend fun onCalendarMonthChanged(month: Instant) {
        if (monthCacheKey(month) in cachedMonthKeys) {
            _activeDaysInMonth.value = computeActiveDays(month)
        } else {
            perform { loadMonth(month) }
        }
    }

    // MARK: - Private

    private suspend fun perform(
        onFailure: (Exception) -> Unit = { alertItem.value = unknownErrorAlertItem(it) },
        block: suspend () -> Unit,
    ) {
        try {
            block()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.DASHBOARD)
            onFailure(error)
        }
    }

    private fun unknownErrorAlertItem(error: Throwable): AlertItem =
        if (error.isFirestoreUnreachable) {
            AlertItem(titleRes = R.string.common_error_offline, messageRes = R.string.common_error_offline_message)
        } else {
            AlertItem(titleRes = R.string.common_error_unknown, messageRes = R.string.common_error_unknown_message)
        }

    private suspend fun refreshMealTypes() {
        var types = fetchMealTypes()
        if (types.isEmpty() && confirmMealTypesEmpty()) {
            types = setupDefaultMeals()
        }
        mealTypes.value = types
    }

    private fun advanceSelectedDayIfNeeded() {
        if (!isViewingToday) return
        if (selectedDay.value.isSameDay(Instant.now())) return
        selectedDay.value = Instant.now()
    }

    private suspend fun loadFoods(date: Instant) {
        if (monthCacheKey(date) in cachedMonthKeys) {
            foodsConsumed.value = foodsFromCache(date)
            _activeDaysInMonth.value = computeActiveDays(date)
        } else {
            perform {
                loadMonth(date)
                foodsConsumed.value = foodsFromCache(date)
            }
        }
    }

    private suspend fun loadMonth(date: Instant) {
        val foods = fetchFoodsConsumedForMonth(date)
        populateCache(foods, date)
        _activeDaysInMonth.value = computeActiveDays(date)
    }

    private fun monthCacheKey(date: Instant): String = date.formatCacheKey("yyyy-MM")

    private fun dayCacheKey(date: Instant): String = date.formatCacheKey("yyyy-MM-dd")

    private fun populateCache(foods: List<FoodConsumedDomain>, month: Instant) {
        val key = monthCacheKey(month)
        monthCache = monthCache.filterKeys { !it.startsWith(key) }.toMutableMap()
        cachedMonthKeys += key
        for ((dayKey, dayFoods) in foods.groupBy { dayCacheKey(it.date) }) {
            monthCache[dayKey] = monthCache[dayKey].orEmpty() + dayFoods
        }
    }

    private fun foodsFromCache(date: Instant): List<FoodConsumedDomain> = monthCache[dayCacheKey(date)].orEmpty()

    private fun computeActiveDays(month: Instant): Set<Int> {
        val key = monthCacheKey(month)
        val result = mutableSetOf<Int>()
        for ((dayKey, foods) in monthCache) {
            if (!dayKey.startsWith(key) || foods.isEmpty()) continue
            val parts = dayKey.split("-")
            val day = parts.getOrNull(2)?.toIntOrNull()
            if (parts.size == 3 && day != null) {
                result += day
            }
        }
        return result
    }

    private fun invalidateCache(date: Instant) {
        val key = monthCacheKey(date)
        monthCache = monthCache.filterKeys { !it.startsWith(key) }.toMutableMap()
        cachedMonthKeys -= key
    }
}
